java-checks/src/main/java/org/sonar/java/checks/serialization/SerializableFieldInSerializableClassCheck.java

```diff
@@ -54,32 +65,65 @@ public void visitNode(Tree tree) {
     if (isSerializable(classTree)
       && !SerializableContract.hasSpecialHandlingSerializationMethods(classTree)
       && !classTree.symbol().type().isSubtypeOf("javax.servlet.http.HttpServlet")) {
-      for (Tree member : classTree.members()) {
-        if (member.is(Tree.Kind.VARIABLE)) {
-          checkVariableMember((VariableTree) member);
-        }
-      }
+
+      Set<String> constructorInjectedParams = getConstructorInjectedFields(classTree);
+      Set<String> setterInjectedParams = getSetterInjectedFields(classTree);
+
+      classTree.members().stream()
+        .filter(member -> member.is(Tree.Kind.VARIABLE))
+        .map(VariableTree.class::cast)
+        .filter(variableTree -> !(isExcluded(variableTree) ||
+          constructorInjectedParams.contains(variableTree.simpleName().name()) ||
+          setterInjectedParams.contains(variableTree.simpleName().name())))
+        .forEach(this::checkVariableMember);
     }
   }
 
+  private static Set<String> getSetterInjectedFields(ClassTree classTree) {
+    return classTree.members().stream()
+      .filter(member -> member.is(Tree.Kind.METHOD))
+      .map(MethodTree.class::cast)
+      .filter(methodTree -> isAnnotatedWith(methodTree.symbol().metadata(), JAVAX_INJECT))
+      .map(methodTree -> methodTree.simpleName().name())
+      .filter(name -> name.startsWith("set"))
+      .map(name -> name.substring(3).toLowerCase(Locale.ROOT))
+      .collect(Collectors.toSet());
+  }
+
+  private static Set<String> getConstructorInjectedFields(ClassTree classTree) {
+    Set<String> fields = classTree.members().stream()
+      .filter(member -> member.is(Tree.Kind.VARIABLE))
+      .map(VariableTree.class::cast)
+      .map(variableTree -> variableTree.simpleName().name())
+      .collect(Collectors.toSet());
+
+    AssignmentsVisitor assignmentsVisitor = new AssignmentsVisitor(fields);
+    
+    classTree.members().stream()
+      .filter(member -> member.is(Tree.Kind.CONSTRUCTOR))
+      .map(MethodTree.class::cast)
+      .filter(methodTree -> isAnnotatedWith(methodTree.symbol().metadata(), JAVAX_INJECT))
+      .forEach(methodTree -> methodTree.accept(assignmentsVisitor));
+    
+    return assignmentsVisitor.getUsedFields();
```

Comment:
This should work fine, I would just add a test when you have multiple constructors reaching the `forEach`, to make sure we are not losing anything across different visits.

Reply:
Actually this is impossible situation for compiling example, cause only one constructor can be annotated with <code>@Inject</code> But I added it in non-compiling example